include/swift/ABI/Metadata.h

```diff
@@ -804,14 +804,39 @@ struct TargetVTableDescriptorHeader {
   /// entries occupy in instantiated class metadata.
   uint32_t VTableSize;
 
-  uint32_t getVTableOffset(const TargetClassMetadata<Runtime> *metadata) const {
-    const auto *description = metadata->getDescription();
-    if (description->hasResilientSuperclass())
-      return metadata->Superclass->getSizeInWords() + VTableOffset;
+  uint32_t getVTableOffset(const TargetClassDescriptor<Runtime> *description) const {
+    if (description->hasResilientSuperclass()) {
+      auto bounds = description->getMetadataBounds();
+      return (bounds.ImmediateMembersOffset / sizeof(StoredPointer)
+              + VTableOffset);
+    }
+
     return VTableOffset;
   }
 };
 
+template <typename Runtime>
+struct TargetMethodOverrideDescriptor {
+  /// The class containing the base method.
+  TargetRelativeIndirectablePointer<Runtime, TargetClassDescriptor<Runtime>> Class;
+
+  /// The base method.
+  TargetRelativeIndirectablePointer<Runtime, TargetMethodDescriptor<Runtime>> Method;
+
+  /// The implementation of the override.
+  TargetRelativeDirectPointer<Runtime, void, /*Nullable=*/true> Impl;
+};
+
+/// Header for a class vtable override descriptor. This is a variable-sized
+/// structure that provides implementations for overrides of methods defined
+/// in superclasses.
+template <typename Runtime>
+struct TargetOverrideTableHeader {
+  /// The number of MethodOverrideDescriptor records following the vtable
+  /// override header in the class's nominal type descriptor.
+  uint32_t NumEntries;
```

Comment:
Can we reserve another uint32_t here? We need it for alignment anyway on 64-bit platforms, so we might as well make that explicit.

Alternately, is it really a problem to put another 32-bit field into every class descriptor?

Reply:
I don't think we need to align these descriptors on pointer-size boundaries (as John pointed out I'm setting the alignment to pointer size in a few places but that's incorrect). They only contain relative pointers so they're 32-bit aligned. What were you thinking of using the other 32 bits for?

Reply:
If we want to save space for random expansion — and I'm not saying we definitely need to, but maybe Jordan has something in mind — we should prefer taking bits from the count instead of adding flags.

I don't think we need to support more than 64K new methods or overrides in a single class.  That would be an awful lot of code.

Reply:
Whoops, I forgot that we compacted relative pointers down to 32 bits. That makes it all less important.